### ISSUE TITLE
Spring HATEOAS and custom (vendor-specific) media-types

### ISSUE THREAD

**Issue by vivin** [spring-projects/spring-hateoas#263]
I'm using Spring HATEOAS with Spring Boot for a service. Originally we simply had the endpoints returning `application/json`. Links and everything were created properly by Spring HATEOAS, but I noticed that the `links` property was an array of maps. Instead, of simply being a map where the `rel`s are keys. I was able to fix that by adding `@EnableHypermediaSupport(type=EnableHypermediaSupport.HypermediaType.HAL)`.

My next step was to support semantic media-types (Level 3 of Richardson Maturity Model). I was able to create custom media-type converters for my custom media types by extending `MappingJackson2HttpMessageConverter`, and then adding the appropriate `produces` and `consumes` values on my `@RequestMapping` annotations. 

However, I noticed that now the links are back to the old format where they are an array of hashes. I imagine this is because the converters are extending `MappingJackson2HttpMessageConverter`. I looked at the source code for Spring HATEOAS and I see that the converter that handle `application/hal+json` is being defined in `HyperMediaSupportBeanDefinitionRegistrar`:

```
private List<HttpMessageConverter<?>> potentiallyRegisterModule(List<HttpMessageConverter<?>> converters) {
    ...

    CurieProvider curieProvider = getCurieProvider(beanFactory);
    RelProvider relProvider = beanFactory.getBean(DELEGATING_REL_PROVIDER_BEAN_NAME, RelProvider.class);
    ObjectMapper halObjectMapper = beanFactory.getBean(HAL_OBJECT_MAPPER_BEAN_NAME, ObjectMapper.class);

    halObjectMapper.registerModule(new Jackson2HalModule());
    halObjectMapper.setHandlerInstantiator(new Jackson2HalModule.HalHandlerInstantiator(relProvider, curieProvider));

    MappingJackson2HttpMessageConverter halConverter = new MappingJackson2HttpMessageConverter();
    halConverter.setSupportedMediaTypes(Arrays.asList(HAL_JSON)); //HAL_JSON is just a MediaType instance for application/hal+json
    halConverter.setObjectMapper(halObjectMapper);

    ...
}
```

How can I ensure that HAL-type links work with my custom media-types? Will I have to duplicate this code inside each of my converters?

**Comment by kewne** [comment]
@vivin, correct me if my interpretation of the discussion is wrong (I know it's late and you've probably fixed your original problem but the issue is still open): your requirement of custom media types seems to revolve about assigning a media type to a given domain entity such that, for example, an order would be a response with media type `application/vnd.order.v1` (or similar, I forget the exact format).

While I agree that supporting more media types and, particularly, removing the current mandatory usage of HAL, I'd like to challenge the notion that Level 3 in the Richardson REST Maturity Model **implies** usage of custom media types, especially to represent domain entities: the way I understand his QCon presentation, he describes custom media types as something that's somewhat tied to user agents (hope I'm using the term correctly) and the context of the data.
He even uses ATOM (which represents news feeds to be used be feed readers) and HTML (which represents a "printable" document used by browsers)

What you seem to require is link semantics, meaning that a link of type "order" is a link to an object that contains order information.
However, **media type is merely a way of representing that order resource**, which could be represented as `HAL+JSON`, `HAL+XML` or `ATOM`;
what determines the semantics of a resource represented as a media type is the link's `rel` attribute, which can be extended by using CURIE.

Sorry for the long comment and let me know if I totally missed the point.

**Comment by vivin** [comment]
@kewne No worries! Yeah, I agree that Level 3 doesn't **mandate** custom media-types; it only mandates the used of **semantically-appropriate** media-types, in the sense that the user-agent has the ability to specify the type of representation that it wants for a particular resource. So the media-type for a resource basically `application/<what-resource-means>+<what-resource-looks-like>`. 

I think you helped clarify things for me :). I was going about the whole thing wrong anyway because whether something uses HAL or not would (should) be documented for the media-type. So a media-type for a resource that has HAL links is necessarily different from one that does not. HAL describes the linking semantics of the resource, so a representation of the same resource without HAL has a different meaning, in the sense that it may convey linking semantics in a completely different way or not at all. I think that was the root of my mistake - I assumed that the HAL portion changed the *way the resource was represented* when really what it does is provide additional semantics for *the way the resource is linked to other resources* - now that information could be represented in different ways like you said: as XML or JSON, or some other format.

**Comment by shuraa** [comment]
So, there is no simple way to register custom media types with HAL support?

**Comment by elnur** [comment]
TBH, Spring HATEOAS seems to be abandoned most of the time. It gets so little attention from the contributors. Have to patch it up all over the place. 😞

**Comment by elnur** [comment]
So here's my current hack to solve the problem. It only makes sense if you only emit HAL and don't support other representations.

First, I enable hypermedia support *explicitly*: 

```java
@EnableHypermediaSupport(type = HAL)
```

That fixes the problem in no time but introduces other problems like breaking a lot of nice support and configurability of Jackson. For instance, you can't configure stuff like the following in your `application.yml` file anymore:

```yaml
spring:
  jackson:
    serialization:
      write_dates_as_timestamps: false
```

Now, to bring all those awesome features back, I have a configuration class that uses the default object mapper builder to configure the object mapper created by Spring HATEOAS:

```java
@Configuration
class JacksonConfig {
    @Autowired
    private ObjectMapper objectMapper;
    @Autowired
    private Jackson2ObjectMapperBuilder objectMapperBuilder;

    @Bean
    ObjectMapper objectMapper() {
        objectMapperBuilder.configure(objectMapper);
        return objectMapper;
    }
}
```

This is the simplest hack I've seen so far. If someone comes up with a simpler one, I'll be glad to switch to that.

**Comment by jack-harrison** [comment]
All workarounds appear to be ugly. In our case, this was the least invasive option, by simply adding an additional supported media type to the existing message converter. No custom message converters, no overriding of object mappers. Just a simple self-contained class which can be removed if this issue gets resolved one day.

```java
@Component
public class HalMediaTypeEnabler {
    private static final MediaType CUSTOM_MEDIA_TYPE = new MediaType("application", "*+hal+json");
    private final RequestMappingHandlerAdapter requestMappingHandlerAdapter;

    @Autowired
    HalMediaTypeEnabler(RequestMappingHandlerAdapter requestMappingHandlerAdapter) {
        this.requestMappingHandlerAdapter = requestMappingHandlerAdapter;
    }

    @PostConstruct
    public void enableVndHalJson() {
        for (HttpMessageConverter<?> converter : requestMappingHandlerAdapter.getMessageConverters()) {
            if (converter instanceof MappingJackson2HttpMessageConverter && converter.getSupportedMediaTypes().contains(HAL_JSON)) {
                MappingJackson2HttpMessageConverter messageConverter = (MappingJackson2HttpMessageConverter) converter;
                messageConverter.setSupportedMediaTypes(Arrays.asList(HAL_JSON, CUSTOM_MEDIA_TYPE));
            }
        }
    }
}
```

**Comment by jkubrynski** [comment]
Any progress on this issue?

**Comment by gregturn** [comment]
I have spent quite a bit of time upgrading Spring Data REST to use the new Spring HATEAOS Affordances API. This means SDR being able to product not only HAL but HAL-FORMS and Collection+JSON documents.

So some of these other issues have not received as much love and attention.

**Comment by vpavic** [comment]
Are there any updates on this issue? We're also using a custom media type that's based on `application/hal+json` and were forced to resort to similar workaround as outlined by @jack-harrison.

Our concrete solution was to do it via `BeanPostProcessor`:

```java
public class RequestMappingHandlerAdapterPostProcessor implements BeanPostProcessor {

    @Override
    public Object postProcessBeforeInitialization(@NonNull Object bean, String beanName) {
        if (bean instanceof RequestMappingHandlerAdapter) {
            RequestMappingHandlerAdapter adapter = (RequestMappingHandlerAdapter) bean;
            for (HttpMessageConverter<?> messageConverter : adapter.getMessageConverters()) {
                if (isHateoasEnabledMessageConverter(messageConverter)) {
                    configureMediaTypes((MappingJackson2HttpMessageConverter) messageConverter);
                }
            }
        }
        return bean;
    }

    private static boolean isHateoasEnabledMessageConverter(HttpMessageConverter<?> messageConverter) {
        return (messageConverter instanceof MappingJackson2HttpMessageConverter)
                && messageConverter.getSupportedMediaTypes().contains(MediaTypes.HAL_JSON);
    }

    private static void configureMediaTypes(MappingJackson2HttpMessageConverter messageConverter) {
        messageConverter.setSupportedMediaTypes(Collections.singletonList(MyMediaTypes.VND_JSON));
    }

}
```

**Comment by gregturn** [comment]
I’m actually working on a new feature (https://github.com/spring-projects/spring-hateoas/tree/feature/new-mediatypes) to register your own media type.

**Comment by vpavic** [comment]
That great to hear @gregturn - is this expected to make it into `1.0.0.M1`?

**Comment by gregturn** [comment]
@vpavic We're knocking 'em out as we can, so I'm not really sure.

**Comment by gregturn** [comment]
Superceded and resolved via #833.
